### PR TITLE
feat(subscriber): filter consumed messages on type

### PR DESCRIPTION
**Pull request by benniekrijger**

### CODE REVIEW

File: src/main/scala/nl/tradecloud/kafka/KafkaSubscriber.scala [hunk 1]
```diff
@@ -62,20 +63,41 @@ class KafkaSubscriber(
...
-          log.debug("Committed offset, msg={}", msg)
+          log.debug("committed offset, msg={}", msg)
           result
         }
       }
   }
 
-  def atLeastOnce(flow: Flow[KafkaMessage, CommittableOffset, _]): Future[Done] = {
+  val commitSink: Sink[CommittableOffset, _] = {
+    commitFlow.to(Sink.ignore)
```
Comment (marcmatt) [inline src/main/scala/nl/tradecloud/kafka/KafkaSubscriber.scala]: Could you explain how this indirection works?

File: src/test/scala/nl/tradecloud/kafka/KafkaSubscriberSpec.scala [hunk 2]
```diff
@@ -105,6 +106,33 @@ class KafkaSubscriberSpec extends TestKit(ActorSystem("KafkaSubscriberSpec", Con
...
+        topics = Set("test_topic_5"),
+        system = system,
+        configurationProperties = Seq(ConsumerConfig.AUTO_OFFSET_RESET_CONFIG -> "earliest")
+      )
+
+      subscriber.atLeastOnce[String](
+        Flow[KafkaMessage[String]].map { msg =>
+          receiverProbe.ref ! msg.msg
+
+          msg.offset
```
Comment (marcmatt) [inline src/test/scala/nl/tradecloud/kafka/KafkaSubscriberSpec.scala]: What does  msg.offset do?

File: src/test/scala/nl/tradecloud/kafka/KafkaSubscriberSpec.scala [hunk 3]
```diff
@@ -105,6 +106,33 @@ class KafkaSubscriberSpec extends TestKit(ActorSystem("KafkaSubscriberSpec", Con
...
+        }
+      )
+
+      publisher.publish(
+        topic = "test_topic_5",
+        msg = Done
+      )
+      receiverProbe.expectNoMessage(defaultNegativeTimeout)
```
Comment (marcmatt) [inline src/test/scala/nl/tradecloud/kafka/KafkaSubscriberSpec.scala]: Can we test the discarded message is committed?

File: src/test/scala/nl/tradecloud/kafka/KafkaSubscriberSpec.scala (hunk 2 above)
Comment (benniekrijger) [reply src/test/scala/nl/tradecloud/kafka/KafkaSubscriberSpec.scala]: The flow argument passed to atLeastOnce function, is of type Flow[KafkaMessage, CommittableOffset, _], so it should output the offset of type CommittableOffset

File: src/test/scala/nl/tradecloud/kafka/KafkaSubscriberSpec.scala (hunk 3 above)
Comment (benniekrijger) [reply src/test/scala/nl/tradecloud/kafka/KafkaSubscriberSpec.scala]: We can test if the commitScaladsl function on the CommittableOffset is called, the actual commit should be (and is) tested at the kafka streams library.

File: src/main/scala/nl/tradecloud/kafka/KafkaSubscriber.scala (hunk 1 above)
Comment (benniekrijger) [reply src/main/scala/nl/tradecloud/kafka/KafkaSubscriber.scala]: In order to use the directTo flow shape we need a commit Sink. This create a commit sink. We can ignore the result.

File: src/test/scala/nl/tradecloud/kafka/KafkaSubscriberSpec.scala [hunk 4]
```diff
@@ -105,6 +106,68 @@ class KafkaSubscriberSpec extends TestKit(ActorSystem("KafkaSubscriberSpec", Con
...
+      val control = subscriber
+        .consumerStream(
+          Flow[KafkaMessage[String]].map { msg =>
+            receiverProbe.ref ! msg.msg
+
+            msg.offset
+          }
+        )
+        .map { _ =>
+          committedCount += 1
```
Comment (marcmatt) [inline src/test/scala/nl/tradecloud/kafka/KafkaSubscriberSpec.scala]: Could you please explain me (by Zoom) how this tests the commit of the divertTo flow

Comment (benniekrijger) [reply src/test/scala/nl/tradecloud/kafka/KafkaSubscriberSpec.scala]: The consumerStream in the KafkaSubscriber contains the Source and all flow shapes which are run when a Sink is connected. The last flow shape in the consumerStream is the commitFlow, committing is performed in this flow. 

The test adds a map flow and a Sink to this stream and if an element reaches this map flow, it must have gone through the commitFlow. Thereby proving the commitFlow is passed by each element, which meets the consumerStream conditions (type check etc).

**Pull request merged**